src/dbnode/integration/wide_query_test.go

```diff
@@ -0,0 +1,343 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package integration
+
+import (
+	"fmt"
+	"io/ioutil"
+	"runtime"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/sharding"
+	"github.com/m3db/m3/src/dbnode/storage"
+	"github.com/m3db/m3/src/dbnode/storage/index"
+	"github.com/m3db/m3/src/m3ninx/idx"
+	xclock "github.com/m3db/m3/src/x/clock"
+	"github.com/m3db/m3/src/x/context"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/ident"
+	xtest "github.com/m3db/m3/src/x/test"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+type shardedIndexChecksum struct {
+	shard     uint32
+	checksums []ident.IndexChecksum
+}
+
+// buildExpectedChecksumsByShard sorts the given IDs into ascending shard order,
+// applying given shard filters.
+func buildExpectedChecksumsByShard(
+	ids []string,
+	allowedShards []uint32,
+	shardSet sharding.ShardSet,
+) []ident.IndexChecksum {
+	shardedChecksums := make([]shardedIndexChecksum, 0, len(ids))
+	count := 0
+	for i, id := range ids {
+		checksum := ident.IndexChecksum{ID: []byte(id), Checksum: int64(i)}
+		shard := shardSet.Lookup(ident.BytesID([]byte(id)))
+
+		if len(allowedShards) > 0 {
+			shardInUse := false
+			for _, allowed := range allowedShards {
+				if allowed == shard {
+					shardInUse = true
+					break
+				}
+			}
+
+			if !shardInUse {
+				continue
+			}
+		}
+
+		found := false
+		for idx, sharded := range shardedChecksums {
+			if shard != sharded.shard {
+				continue
+			}
+
+			found = true
+			shardedChecksums[idx].checksums = append(sharded.checksums, checksum)
+			break
+		}
+
+		if found {
+			continue
+		}
+
+		count++
+		shardedChecksums = append(shardedChecksums, shardedIndexChecksum{
+			shard:     shard,
+			checksums: []ident.IndexChecksum{checksum},
+		})
+	}
+
+	sort.Slice(shardedChecksums, func(i, j int) bool {
+		return shardedChecksums[i].shard < shardedChecksums[j].shard
+	})
+
+	checksums := make([]ident.IndexChecksum, 0, count)
```

Comment:
Same argument as above. Way less symbols for the eyes to parse:
```suggestion
	var checksums []ident.IndexChecksum
```